nilearn/glm/first_level/first_level.py

```diff
@@ -959,107 +950,387 @@ def first_level_from_bids(dataset_path, task_label, space_label=None,
+                raise ValueError(f"{len(events)} confounds.tsv files found "
+                                 f"for {len(imgs)} bold files. "
+                                 "Same number of confound files as "
+                                 "the number of runs is expected")
             confounds = [pd.read_csv(c, sep='\t', index_col=None)
                          for c in confounds]
+
             models_confounds.append(confounds)
 
     return models, models_run_imgs, models_events, models_confounds
+
+
+def _supported_bids_filter() -> dict[str, list[str]]:
+    """Return a dictionary of BIDS entities.
+
+    Note that:
+
+    - this only contains the entities for functional data
+    - the entities `sub` and `ses` are not included
+
+    Returns
+    -------
+    Dictionary of raw and derivatives entities : dict[str, list[str]]
+
+    """
+    return {"raw": ['acq',
+                    'ce',
+                    'rec',
+                    'dir',
+                    'run',
+                    'echo',
+                    'part'],
+            "derivatives": ['res', 'den', 'desc']
+            }
+
+
+def _validate_args_first_level_from_bids(dataset_path: str,
+                                         task_label: str,
+                                         space_label: str | None,
+                                         img_filters: list[Tuple[str, str]],
+                                         derivatives_path: str,
+                                         ) -> None:
+    """Check type and value of arguments of first_level_from_bids.
+
+    Check that:
+        - dataset_path is a string and exists
+        - derivatives_path exists
+        - task_label and space_label are strings
+        - img_filters is a list of tuples of strings
+          and all filters are valid bids entities.
+
+    Parameters
+    ----------
+    dataset_path : :obj:`str`
+        Fullpath of the BIDS dataset root folder.
+
+    task_label : :obj:`str`
+        Task_label as specified in the file names like _task-<task_label>_.
+
+    space_label : :obj:`str`
+        Specifies the space label of the preprocessed bold.nii images.
+        As they are specified in the file names like _space-<space_label>_.
+
+    img_filters : :obj:`list` of :obj:`tuples` (str, str)
+        Filters are of the form (field, label).
+        Only one filter per field allowed.
+
+    derivatives_path : :obj:`str`
+        Fullpath of the BIDS dataset derivative folder.
+
+    """
+    if not isinstance(dataset_path, str):
+        raise TypeError(
+            "dataset_path must be a string, "
+            f"instead {type(dataset_path)} was given"
+        )
+    if not os.path.exists(dataset_path):
+        raise ValueError(f'given path do not exist: {dataset_path}')
+
+    if not os.path.exists(derivatives_path):
+        raise ValueError('derivatives folder does not exist in given dataset')
+
+    if not isinstance(task_label, str):
```

Comment:
TODO: BIDS labels are not just strings but strings that can only contain alphanumeric  characters.

Reply:
done